### PR TITLE
👨‍🏭flatten page metadata in output model, template raw model part-2

### PR DESCRIPTION
**Pull request by OsmondJiang**
this is anthor change to make our page model more close to raw model for template.

flatten metadata in page model.

### CODE REVIEW

File: src/docfx/build/page/PageModel.cs [hunk 1]
```diff
@@ -48,10 +49,11 @@ public class PageModel
...
         public string Gitcommit { get; set; }
 
-        public FileMetadata Metadata { get; set; }
-
         public bool Bilingual { get; set; }
 
         public List<string> Monikers { get; set; }
+
+        [JsonExtensionData]
+        public JObject Metadata { get; set; }
```
Comment (yufeih) [inline src/docfx/build/page/PageModel.cs]:
```suggestion
        public JObject ExtensionData { get; set; }
```

File: src/docfx/build/page/BuildPage.cs [hunk 2]
```diff
@@ -20,18 +20,18 @@ internal static class BuildPage
...
+                var (breadcrumbError, breadcrumbPath, _) = context.DependencyResolver.ResolveLink(fileMetadata.BreadcrumbPath, file, file, buildChild);
                 errors.AddIfNotNull(breadcrumbError);
-                metadata.BreadcrumbPath.Value = breadcrumbPath;
+                fileMetadata.BreadcrumbPath.Value = breadcrumbPath;
             }
 
             model.SchemaType = schema.Name;
             model.Locale = file.Docset.Locale;
-            model.Metadata = metadata;
+            model.Metadata = context.MetadataProvider.GetPageMetadata(fileMetadata);
```
Comment (yufeih) [inline src/docfx/build/page/BuildPage.cs]:
```suggestion
            model.Metadata = fileMetadata;
```

File: src/docfx/build/page/BuildPage.cs [hunk 3]
```diff
@@ -40,7 +40,7 @@ internal static class BuildPage
             (model.ContentGitUrl, model.OriginalContentGitUrl, model.OriginalContentGitUrlTemplate, model.Gitcommit) = context.ContributionProvider.GetGitUrls(file);
 
             List<Error> contributorErrors;
-            (contributorErrors, model.Author, model.Contributors, model.UpdatedAt) = await context.ContributionProvider.GetAuthorAndContributors(file, metadata.Author);
+            (contributorErrors, model.Author, model.Contributors, model.UpdatedAt) = await context.ContributionProvider.GetAuthorAndContributors(file, fileMetadata.Author);
```
Comment (yufeih) [inline src/docfx/build/page/BuildPage.cs]: consider rename `model.Author` --> `model.AuthorInfo`

File: src/docfx/build/page/BuildPage.cs (hunk 2 above)
Comment (OsmondJiang) [reply src/docfx/build/page/BuildPage.cs]: JsonUtility.ToJObject(fileMetadata)

File: src/docfx/build/page/BuildPage.cs (hunk 3 above)
Comment (OsmondJiang) [reply src/docfx/build/page/BuildPage.cs]: There will be a lot of tests need to be change :(

File: src/docfx/template/TemplateEngine.cs [hunk 4]
```diff
@@ -211,14 +211,14 @@ public JObject CreateRawMetadata(PageModel pageModel, Document file)
...
                 {
-                    rawMetadata["_op_gitContributorInformation"]["author"] = ToJObject(pageModel.Author);
+                    rawMetadata["_op_gitContributorInformation"]["author"] = ToJObject(pageModel.AuthorInfo);
                 }
             }
 
-            if (!string.IsNullOrEmpty(pageModel.Author?.Name))
-                rawMetadata["author"] = pageModel.Author?.Name;
+            if (!string.IsNullOrEmpty(pageModel.AuthorInfo?.Name))
+                rawMetadata["author"] = pageModel.AuthorInfo?.Name;
```
Comment (yufeih) [inline src/docfx/template/TemplateEngine.cs]: remove `author_info`

File: src/docfx/build/page/OutputPageModel.cs [hunk 5]
```diff
@@ -4,25 +4,22 @@
...
 using System.Collections.Generic;
 using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
 using Newtonsoft.Json.Serialization;
 
 namespace Microsoft.Docs.Build
 {
     [JsonObject(NamingStrategyType = typeof(SnakeCaseNamingStrategy))]
-    public class PageModel
+    public class OutputPageModel : InputMetadata
```
Comment (yufeih) [inline src/docfx/build/page/OutputPageModel.cs]: `OutputMetadata` ?

Comment (OsmondJiang) [reply src/docfx/build/page/OutputPageModel.cs]: I prefer OutputModel

File: src/docfx/build/page/OutputModel.cs [hunk 6]
```diff
@@ -4,25 +4,22 @@
...
 using System.Collections.Generic;
 using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
 using Newtonsoft.Json.Serialization;
 
 namespace Microsoft.Docs.Build
 {
     [JsonObject(NamingStrategyType = typeof(SnakeCaseNamingStrategy))]
-    public class PageModel
+    public class OutputModel : InputMetadata
```
Comment (superyyrrzz) [inline src/docfx/build/page/OutputModel.cs]: Not all fields in InputMetadata is in OutputModel, e.g. monikerRange. Seems clearer to separate them.

Comment (OsmondJiang) [reply src/docfx/build/page/OutputModel.cs]: The previous behavior actually is also put all input metadata into output model, this is just another way to do that.

I will leave this problem into my later PRs. @superyyrrzz

Comment (yufeih) [reply src/docfx/build/page/OutputModel.cs]: > Not all fields in InputMetadata is in OutputModel, e.g. monikerRange. Seems clearer to separate them.

That one is currently solved by `public bool ShouldSerializeMonikerRange() => false;`

**Pull request merged**